File: 1008 Construct Binary Search Tree from Preorder Traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def bstFromPreorder(self, preorder: List[int]) -> TreeNode:
        if not preorder: return None
        return self.build(preorder, 0, len(preorder) - 1)

    def build(self, preorder: List[int], l, r) -> TreeNode:
        if l > r: return None
        root = TreeNode(preorder[l])
        i = l + 1
        while i <= r and preorder[i] < preorder[l]:
            i += 1
        root.left = self.build(preorder, l + 1, i - 1)
        root.right = self.build(preorder, i, r)
        return root

''' O(N) Solution from lee215
class Solution:
    i = 0
    def bstFromPreorder(self, A, bound=float('inf')):
        if self.i == len(A) or A[self.i] > bound:
            return None
        root = TreeNode(A[self.i])
        self.i += 1
        root.left = self.bstFromPreorder(A, root.val)
        root.right = self.bstFromPreorder(A, bound)
        return root
'''

File: 1008 Construct Binary Search Tree from Preorder Traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public TreeNode bstFromPreorder(int[] preorder) {
        if (preorder == null || preorder.length == 0) return null;
        return build(preorder, 0, preorder.length - 1);
    }
    public TreeNode build(int[] preorder, int l, int r) {
        if (l > r) return null;
        TreeNode root = new TreeNode(preorder[l]);
        int i = l + 1;
        while (i <= r && preorder[i] < preorder[l]) ++i;
        root.left = build(preorder, l + 1, i - 1);
        root.right = build(preorder, i, r);
        return root;
    }
}

/* O(N) Solution from lee215

class Solution {
    int i = 0;
    public TreeNode bstFromPreorder(int[] preorder) {
        if (preorder == null || preorder.length == 0) return null;
        return build(preorder, Integer.MAX_VALUE);
    }
    public TreeNode build(int[] preorder, int bound) {
        if (i == preorder.length || preorder[i] > bound) return null;
        TreeNode root = new TreeNode(preorder[i++]);
        root.left = build(preorder, root.val);
        root.right = build(preorder, bound);
        return root;
    }
}

*/